java/fury-core/src/main/java/org/apache/fury/reflect/TypeRef.java

```diff
@@ -0,0 +1,828 @@
+      return true;
+    }
+    if (formalType instanceof WildcardType) {
+      WildcardType your = canonicalizeWildcardType(declaration, (WildcardType) formalType);
+      // if "formalType" is <? extends Foo>, "this" can be:
+      // Foo, SubFoo, <? extends Foo>, <? extends SubFoo>, <T extends Foo> or
+      // <T extends SubFoo>.
+      // if "formalType" is <? super Foo>, "this" can be:
+      // Foo, SuperFoo, <? super Foo> or <? super SuperFoo>.
+      for (Type bound : your.getUpperBounds()) {
+        if (!of(bound).isSupertypeOf(type)) {
+          return false;
+        }
+      }
+      for (Type bound : your.getLowerBounds()) {
+        if (!of(bound).isSupertypeOf(type)) {
+          return false;
+        }
+      }
+      return true;
+    }
+    return canonicalizeWildcardsInType(type).equals(canonicalizeWildcardsInType(formalType));
+  }
+
+  private static WildcardType canonicalizeWildcardType(
+      TypeVariable<?> declaration, WildcardType type) {
+    Type[] declared = declaration.getBounds();
+    List<Type> upperBounds = new ArrayList<>();
+    upperBoundsGenerator:
+    for (Type bound : type.getUpperBounds()) {
+      for (Type declaredType : declared) {
+        if (of(declaredType).isSubtypeOf(bound)) {
+          continue upperBoundsGenerator;
+        }
+      }
+      upperBounds.add(canonicalizeWildcardsInType(bound));
+    }
+    return new Types.WildcardTypeImpl(type.getLowerBounds(), upperBounds.toArray(new Type[0]));
+  }
+
+  private static Type canonicalizeWildcardsInType(Type type) {
+    if (type instanceof ParameterizedType) {
+      ParameterizedType parameterizedType = (ParameterizedType) type;
+      Class<?> rawType = (Class<?>) parameterizedType.getRawType();
+      TypeVariable<?>[] typeVars = rawType.getTypeParameters();
+      Type[] typeArgs = parameterizedType.getActualTypeArguments();
+      for (int i = 0; i < typeArgs.length; i++) {
+        Type typeArg = typeArgs[i];
+        typeArgs[i] =
+            typeArg instanceof WildcardType
+                ? canonicalizeWildcardType(typeVars[i], (WildcardType) typeArg)
+                : canonicalizeWildcardsInType(typeArg);
+      }
+      return new Types.ParameterizedTypeImpl(parameterizedType.getOwnerType(), rawType, typeArgs);
+    }
+    if (type instanceof GenericArrayType) {
+      return Types.newArrayType(
+          canonicalizeWildcardsInType(((GenericArrayType) type).getGenericComponentType()));
+    }
+    return type;
+  }
+
+  private static boolean anyTypeIsSubTypeOf(Type upperBound, Type[] declared) {
+    for (Type declaredType : declared) {
+      if (of(declaredType).isSubtypeOf(upperBound)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private boolean anyRawTypeIsSubclassOf(Class<?> supertype) {
+    return getRawTypes(type).anyMatch(supertype::isAssignableFrom);
+  }
+
+  /** Returns true if this type is a supertype of the given {@code type}. */
+  public final boolean isSupertypeOf(Type type) {
```

Comment:
Could we change to:
 ```java
  public final boolean isSupertypeOf(Type type, boolean generic) {
      if (generic) {
         throw new UnsupportedException(xxx);
      }
```

In this way, the code will be much simpler.

Currently Fury didn't use covariant geneics. Such checks is complicated and increate the maintain cost. If we do need it, we can add it back in future

Reply:
If we don't need the generic covariant  checks. Many methods such as canonicalizeWildcardType can be removed.

Reply:
when generic is noe used ,this can be simplified into :
```java
public final boolean isSupertypeOf(TypeRef<?> type) {
    return getRawType().isAssignableFrom(type.getRawType());
  }
```